refactor(zone.js): remove onProp eventName array to reduce bundle size

JiaLiPassion opened the pull request:
Zone.js supports google closure compiler in advance optimization mode,
to prevent closure compiler modify the `onProperty` name such as `Element.prototype.onclick`,
Zone.js implements the `onProperty` patch logic by declaring all the
event names in source code, this increases the bundle size and also require
updating the event names array to keep the information updated.

Now google closure compiler has the required event names defined in the built-in
externs files, so zone.js can use more simple implementation and decrease the bundle size
of zone.js. (about 3k)

JiaLiPassion committed:
refactor(zone.js): remove onProp eventNames array to reduce the bundle size

Zone.js supports the google closure compiler in the advance optimization mode,
to prevent closure compiler modify the `onProperty` name such as `Element.prototype.onclick`,
Zone.js implements the `onProperty` patch logic by declaring all the
event names in the source code, this increases the bundle size and also require
updating the event names array to keep the information updated.

Now google closure compiler has the required event names defined in the built-in
externs files, so zone.js can use more simple implementation and decrease the bundle size
of zone.js (about 4k).

josephperrott reviewed (approved):
LGTM

Looks like we just need the size tracking to be updated.

AndrewKushnir reviewed (approved):
Great refactoring and almost 4Kb savings! 👍 

Reviewed-for: size-tracking

AndrewKushnir commented:
@JiaLiPassion it looks like there are more places where payload size should be updated, see `test_ivy_aot` CI job output.

Could you please also set the "target" label (would it be safe to land in patch version or only in major)?

JiaLiPassion commented:
@AndrewKushnir , @josephperrott , thank you for the review. I have updated the size and also added the target label.
@AndrewKushnir , could you also run a `presubmit` to check whether it breaks `g3`, I remembered there was an issue about using closure complier advanced optimization failed without these event names.

AndrewKushnir commented:
@JiaLiPassion sure, I've started a [presubmit](http://test/OCL:359172596:BASE:359172611:1614130382899:f0e69ed), will let you know how it goes.

AndrewKushnir commented:
@JiaLiPassion FYI presubmits went well, but I think we'd need to run a global one before merging this change. Thank you.

JiaLiPassion commented:
@AndrewKushnir , got it, thank you!

petebacondarwin reviewed (approved):
Reviewed-for: size-tracking

jelbourn commented on packages/zone.js/lib/browser/api-util.ts:
```diff
@@ -10,9 +10,13 @@ import {globalSources, patchEventPrototype, patchEventTarget, zoneSymbolEventNam
+import {filterProperties} from './property-descriptor';
 
 Zone.__load_patch('util', (global: any, Zone: ZoneType, api: _ZonePrivate) => {
+  const eventNames: string[] = [];
```
I might leave a comment here like
```ts
// Collect native event names by looking at properties
// on the global namespace, e.g. 'onclick'.
```

JiaLiPassion replied:
Got it, thanks, I updated the comment here.

jelbourn commented on packages/zone.js/lib/browser/property-descriptor.ts:
```diff
@@ -263,6 +43,12 @@ export function patchFilteredProperties(
   patchOnProperties(target, filteredProperties, prototype);
 }
 
+function getOnEventNames(target: any) {
```
Could this `any` be something more specific? Even if it's just `object`.

JiaLiPassion replied:
Yeah, it is the prototype, so it can be specific to `object`.

jelbourn commented on packages/zone.js/lib/browser/property-descriptor.ts:
```diff
@@ -271,64 +57,32 @@ export function propertyDescriptorPatch(api: _ZonePrivate, _global: any) {
-  const supportsWebSocket = typeof WebSocket !== 'undefined';
   const ignoreProperties: IgnoreProperty[] = _global['__Zone_ignore_on_properties'];
   // for browsers that we can patch the descriptor:  Chrome & Firefox
+  let patchTargets: any[] = [];
```
Should this be `string[]`?

JiaLiPassion replied:
Yes, I updated the type here, thanks!

jelbourn commented on packages/zone.js/lib/browser/property-descriptor.ts:
```diff
@@ -271,64 +57,32 @@ export function propertyDescriptorPatch(api: _ZonePrivate, _global: any) {
+    'XMLHttpRequest', 'XMLHttpRequestEventTarget', 'IDBIndex', 'IDBRequest', 'IDBOpenDBRequest',
+    'IDBDatabase', 'IDBTransaction', 'IDBCursor', 'WebSocket'
+  ]);
+  patchTargets.forEach(t => {
```
Could this use a `for` loop rather than `.forEach`?

https://github.com/angular/angular/blob/master/docs/CODING_STANDARDS.md#iteration

JiaLiPassion replied:
Got it, I updated the code to `for` loop. Thanks.

jelbourn reviewed (approved):
LGTM

Reviewed-for: size-tracking

mhevery commented:
[presubmit](http://test/OCL:360261758:BASE:360261797:1614633496382:90cf212a)

AndrewKushnir commented:
FYI, adding a cleanup label since there is a merge conflict with the master branch. @JiaLiPassion could you please rebase and resolve the conflict when you get a chance? Thank you.

JiaLiPassion commented:
@AndrewKushnir , got it, I rebased the PR and resolved the conflict, thank you!

AndrewKushnir commented:
@JiaLiPassion thanks for rebasing this PR 👍 I've also started a [global presubmit](https://test.corp.google.com/OCL:359172596:BASE:360831617:1614836992627:90ec4972) (in addition to a subset of tests that were checked before), will let you know how it goes. Thank you.

AndrewKushnir commented:
@JiaLiPassion it looks like there are few affected targets in Google's codebase that would require additional investigation. I'm adding the "blocked" label for now and will share more info once I perform further investigation. Thank you.

JiaLiPassion commented:
@AndrewKushnir , got it, thank you. I will wait for your information.

mhevery commented on packages/zone.js/lib/browser/property-descriptor.ts:
```diff
@@ -263,6 +43,12 @@ export function patchFilteredProperties(
 }
 
+function getOnEventNames(target: Object) {
+  return Object.getOwnPropertyNames(target)
```
I don't believe this function is correct.
`getOnEventNames(XMLHttpRequest)` returns `[]`. The reason for this is that we need to walk the inheritance hierarchy to collect all of the properties. Could you update the code and add some tests which shows that this is the case. 

ALSO: It seems like this function and `for (let key in global) {` has some overlap. Not sure if we can merge the to, but worth investigating.

JiaLiPassion replied:
@mhevery, the current logic is to call `getOnEventNames(XMLHttpRequest.prototype)` which returns `['readystatechange']`, and since `XMLHttpRequest` inherits `XMLHttpRequestEventTarget`, so here https://github.com/angular/angular/pull/40962/files#diff-75568e77504519f5bd9d81b3baacea64225e1f015f030fb0fa767b5b5df69853R79 we also patch the `XMLHttpRequestEventTarget.protptype`, `getOnEventNames(XMLHttpRequestEventTarget.prototype)` returns `["loadstart", "progress", "abort", "error", "load", "timeout", "loadend"]`.

So I believe this is the expect result, could you post more information about the failure in `g3`?

`ALSO: It seems like this function and for (let key in global) { has some overlap. Not sure if we can merge the to, but worth investigating.`

Got it, I will check this one.

mhevery replied:
let me rerun presubmit to see if it helps

alxhub replied:
Can you add a doc comment to this function describing its intention (and specifically, that it only considers the immediate object properties and not any inherited properties).

mhevery reviewed (changes requested):
Failing in g3 due to the comment below.

mhevery commented:
[presubmit](http://test/OCL:360261758:BASE:361853862:1615317487110:e7a84c84)

JiaLiPassion committed:
refactor(zone.js): for legacy browser, still use hard coding eventNames

For legacy browsers, we still use the eventNames array to patch event instead of
using `Object.getOwnPropertyNames()` to keep backward compatibility.

JiaLiPassion commented:
@mhevery , I added another test cases to make sure all `evergreen browsers` are patched correctly, and this PR should patch the same eventNames with the old way.

And I found a breaking change is the `eventNames` array is also used for `legacy browser` patch, and the new way lack of some events, so I put the old eventNames array to the `legacy patch` code here https://github.com/angular/angular/pull/40962/files#diff-4351958690be6f3466945830172d704d8e1d8711ec9eb3fdf41c7abd4b641cd6R110. It will not impact the bundle size of bundle of `evergreen` browsers.

Could you start another `global presubmit`, if still fail, I will create another `debug commit` to check which event is not patched, thank you.

AndrewKushnir commented:
@JiaLiPassion @mhevery FYI I've just tested the most recent changes from this PR with the targets that were failing previously.- the targets are still failing ([internal-only link to test results](https://test.corp.google.com/ui#id=OCL:359172596:BASE:361953196:1615347181700:1a09b9a7)).

JiaLiPassion commented:
@AndrewKushnir , got it, thank you, I will create a debug PR to check which eventName is not patched.

JiaLiPassion commented:
@AndrewKushnir , I created a debug commit, could you run another presubmit? If the error is gone, could you send me the console log where the error occurs without this debug commit? Thank you.

AndrewKushnir commented:
Hi Jia, just ran the tests again, the targets are still failing (and I verified that they work correctly without the change from this PR). Please let me know what we can try next. Thank you.

JiaLiPassion commented:
@AndrewKushnir, it is a little weird that the debug version still failed the `g3`, I updated another `debug` version, could you try the `presubmit` again? Thank you.

JiaLiPassion commented:
@AndrewKushnir, it is a little weird that the debug version still failed the `g3`, I updated another `debug` version, could you try the `presubmit` again? Thank you.

AndrewKushnir commented:
Hi @JiaLiPassion, sure, I can try again. Could you please rebase this PR on top of the most recent version of the master branch and resolve conflicts (so that I can try to sync it again)? Thank you.

JiaLiPassion commented:
@AndrewKushnir , got it, I just rebased the PR, thank you.

AndrewKushnir commented:
Hey @JiaLiPassion, thanks for rebasing.

I've performed additional investigation and found out that the problem was caused by the presence of extra property on the `window` that starts with `on`, something like:
```
window.one_two_three = { ... };
```
as a result, that was picked up by the `getOnEventNames` function as a possible event name. If I filter out that property, everything works as expected.

Is there any protection that we can add (for ex. check the type of the field before patching it)?

Thank you.

JiaLiPassion commented:
Got it, thank you very much for the investigation, I have updated the code to also support onProperty which is not a function, could you run another presubmit?

AndrewKushnir commented:
@JiaLiPassion thanks for the fixes! I've started a new set of tests and will run a broader set of targets tonight as well. I've checked one of the previously failing target and it works correctly now, which is a good sign! Will keep you updated. Thank you.

AndrewKushnir commented:
[Global Presubmit #2](https://test.corp.google.com/OCL:362441781:BASE:362717130:1615663689287:fb41967).

AndrewKushnir commented:
Hi @JiaLiPassion, just a quick update: the most recent change resolved the issue with one target, but there is a couple more failing targets still. We'll need to perform more investigation to find the origin of the problem. Will keep you updated.

JiaLiPassion commented:
@AndrewKushnir , got it, thank you, please tell me if any debug version is needed.

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:427626829:BASE:427626881:1644461610052:6d68a356).

AndrewKushnir commented:
@JiaLiPassion FYI the presubmit went well (all targets are "green").

The change was reviewed and approved previously (almost a year ago), but I'd like to ask @alxhub to take a look as well.

Thank you.

JiaLiPassion commented:
@AndrewKushnir, got it, thank you very much!

alxhub commented on packages/zone.js/lib/browser/property-descriptor-legacy.ts:
```diff
@@ -107,11 +107,218 @@ function canPatchViaPropertyDescriptor(api: _ZonePrivate, _global: any) {
+const formEventNames = ['autocomplete', 'autocompleteerror'];
+const detailEventNames = ['toggle'];
+
+const eventNames = globalEventHandlersEventNames.concat(
```
This is a slightly confusing way to write this. Perhaps:

```typescript
const eventNames = [
  ...globalEventHandlersEventNames,
  ...webglEventNames,
  ...
];
```

would be slightly more readable. Or is this version preferred for code size?

JiaLiPassion replied:
Got it, I updated the code style with `[...]`, thank you.

JiaLiPassion commented:
@alxhub , thank you for the review, I have added the comments and update the code style.

AndrewKushnir commented:
[Global Presubmit #2](https://fusion2.corp.google.com/presubmit/tap/427626829/OCL:427626829:BASE:431596891:1646111892580:557f826c/targets).

AndrewKushnir commented:
FYI, the presubmit is "green", so I'm adding this PR to the merge queue.

AndrewKushnir commented:
Note to the Caretaker: please sync this PR into g3 as a separate change.

jessicajaniuk commented:
This does not merge cleanly with `13.2.x`. I'll update to `target: minor` and land it. Please create a patch PR targeting `13.2.x` and we'll land that one, too.

jessicajaniuk commented:
This PR was merged into the repository by commit 1b85811c00fdf1164649df66f4c20c5da5765e42.
